src/ngx_http_lua_balancer.c

```diff
@@ -489,6 +489,80 @@ ngx_http_lua_ffi_balancer_set_current_peer(ngx_http_request_t *r,
     return NGX_OK;
 }
 
+int
+ngx_http_lua_ffi_balancer_set_timeout(ngx_http_request_t *r,
+    int connect_timeout, int send_timeout, int read_timeout, char **err)
+{
+    ngx_http_lua_ctx_t    *ctx;
+    ngx_http_upstream_t   *u;
+
+    ngx_http_lua_main_conf_t           *lmcf;
+    ngx_http_lua_balancer_peer_data_t  *bp;
+    ngx_http_upstream_conf_t           *ucf;
+
+    if (r == NULL) {
+        *err = "no request found";
+        return NGX_ERROR;
+    }
+
+    u = r->upstream;
+
+    if (u == NULL) {
+        *err = "no upstream found";
+        return NGX_ERROR;
+    }
+
+    ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module);
+    if (ctx == NULL) {
+        *err = "no ctx found";
+        return NGX_ERROR;
+    }
+
+    if ((ctx->context & NGX_HTTP_LUA_CONTEXT_BALANCER) == 0) {
+        *err = "API disabled in the current context";
+        return NGX_ERROR;
+    }
+
+    lmcf = ngx_http_get_module_main_conf(r, ngx_http_lua_module);
+
+    bp = lmcf->balancer_peer_data;
+    if (bp == NULL) {
+        *err = "no upstream peer data found";
+        return NGX_ERROR;
+    }
+
+    if (connect_timeout < 0) {
+        *err = "connect_timeout must be equals or greater than zero";
+        return NGX_ERROR;
+    }
+
+    if (send_timeout < 0) {
+        *err = "send_timeout must be equals or greater than zero";
+        return NGX_ERROR;
+    }
+
+    if (read_timeout < 0) {
+        *err = "read_timeout must be equals or greater than zero";
+        return NGX_ERROR;
+    }
+
+    ucf = ngx_palloc(r->pool, sizeof(ngx_http_upstream_conf_t));
+
+    if (ucf == NULL) {
+        *err = "no memory";
+        return NGX_ERROR;
+    }
+
+    ngx_memcpy(ucf, u->conf, sizeof(ngx_http_upstream_conf_t));
+
+    ucf->connect_timeout = connect_timeout;
+    ucf->send_timeout = send_timeout;
+    ucf->read_timeout = read_timeout;
+
+    u->conf = ucf;
+
+    return NGX_OK;
+}
 
```

Comment:
Style: 2 blank lines needed between function definitions. Better use the `ngx-releng` tool from the `nginx-devel-utils` repos to check your code.